File: ios/Capacitor/Capacitor/UIStatusBarManager+CAPHandleTapAction.m
#import <Capacitor/Capacitor-Swift.h>
#import <objc/message.h>
#import <objc/runtime.h>

@implementation UIStatusBarManager (CAPHandleTapAction)

+ (void)load {
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    Class class = [self class];
    SEL originalSelector = NSSelectorFromString(@"handleTapAction:");
    SEL swizzledSelector = @selector(nofity_handleTapAction:);

    Method originalMethod = class_getInstanceMethod(class, originalSelector);
    Method swizzledMethod = class_getInstanceMethod(class, swizzledSelector);

    BOOL didAddMethod = class_addMethod(class,
                                        originalSelector,
                                        method_getImplementation(swizzledMethod),
                                        method_getTypeEncoding(swizzledMethod));
    if (didAddMethod) {
      class_replaceMethod(class,
                          swizzledSelector,
                          method_getImplementation(originalMethod),
                          method_getTypeEncoding(originalMethod));
    } else {
      method_exchangeImplementations(originalMethod, swizzledMethod);
    }
  });
}

-(void)nofity_handleTapAction:(id)arg1 {
  [[NSNotificationCenter defaultCenter] postNotification:CAPBridge.statusBarTappedNotification];
  [self nofity_handleTapAction:arg1];
}

@end

File: ios/Sources/Capacitor/UIStatusBarManager+CAPHandleTapAction.swift
//
//  UIStatusBarManager+CAPHandleTapAction.swift
//  Capacitor
//
//  Copyright © 2024 Drifty Co. All rights reserved.
//

import Foundation
import UIKit
import ObjectiveC

extension UIStatusBarManager {
    private static let swizzle: Void = {
        let classRef = UIStatusBarManager.self
        let originalSelector = Selector(("handleTapAction:"))
        let swizzledSelector = #selector(UIStatusBarManager.nofity_handleTapAction(_:))

        guard let originalMethod = class_getInstanceMethod(classRef, originalSelector),
              let swizzledMethod = class_getInstanceMethod(classRef, swizzledSelector) else {
            return
        }

        let didAddMethod = class_addMethod(
            classRef,
            originalSelector,
            method_getImplementation(swizzledMethod),
            method_getTypeEncoding(swizzledMethod)
        )

        if didAddMethod {
            class_replaceMethod(
                classRef,
                swizzledSelector,
                method_getImplementation(originalMethod),
                method_getTypeEncoding(originalMethod)
            )
        } else {
            method_exchangeImplementations(originalMethod, swizzledMethod)
        }
    }()

    static func ensureSwizzling() {
        _ = swizzle
    }

    @objc func nofity_handleTapAction(_ arg: Any) {
        NotificationCenter.default.post(
            name: Notification.Name(rawValue: "CapacitorStatusBarTappedNotification"),
            object: nil
        )
        nofity_handleTapAction(arg)
    }
}
